app/controllers/blogs_controller.rb

```diff
@@ -4,6 +4,8 @@ class BlogsController < ApplicationController
   skip_before_action :authenticate_user!, only: %i[index show]
 
   before_action :set_blog, only: %i[show edit update destroy]
+  before_action :ensure_own_blog?, only: %i[edit update destroy]
+  before_action :authenticate_user_when_show_secret_blog, only: %i[show]
```

Comment:
なるほど〜、

- `set_blog` でブログの集合全体から find する
- `set_blog` と別に「リクエストしたユーザが見えてもよいもの」かどうかチェックして、該当しなければ `ActiveRecord::RecordNotFound` を手動で raise する

ということですね。やりたいことができていそうな点はまず good です:+1: その上で、要件通りに動くだけではなく、より Rails らしく保守性が高いとされる書き方に寄せたいです。

まず、Rails の基本原則として、コントローラはなるべく薄く保つ方が良く、モデルやその他のメソッドを呼ぶ程度にした方が好ましいというのがあります。なので、本来はコントローラに書きたくないです。
` ensure_own_blog?` や `authenticate_user_when_show_secret_blog` に書いているロジックからは、**「アクセスできてよいのは公開記事と自分自身の秘密の記事である」というブログ記事に関する業務知識**が浮かび上がってきます。これはコントローラではなく Blog モデルが持っている方がオブジェクト指向的にも自然だったりします。
なので、**Blog モデルの方に「見せて良いものの集合」を返す何かを実装しておいて、ここからはそれに対して find するだけ**、という形が好ましいです〜！

別の観点から言うと、find って引数に渡した id に対応するレコードが見つからないときに ActiveRecord::RecordNotFound を上げてくれるんですよね。

[https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find](https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find)

> Find by id - This can either be a specific id (1), a list of ids (1, 5, 6), or an array of ids ([5, 6, 10]). If one or more records cannot be found for the requested ids, then ActiveRecord::RecordNotFound will be raised.
> 

なので、その意味でもなにかに対して find させて、見つからなければ自然と ActiveRecord::RecordNotFound が発生するように、というのが正攻法だったりします！（つまりここは出題者からの「`find` が raise する ActiveRecord::RecordNotFound を使ってくれぇ〜」というメッセージだったわけですね🙂）

---

それと、異なるアクションを連ねながらインスタンス変数の状態に手を入れていくというのは、往々にして処理の見通しを損なったり、相互に依存してしまって管理しづらくなるのでおすすめしません:warning:
参考: https://speakerdeck.com/shinkufencer/how-to-using-before-action-with-happy-in-rails?slide=11

Reply:
@kfukai23 
ありがとうございます！とても詳しい説明で勉強になりました。
befere_action で変数の設定をしていくと辛みが増していくのが想像できました。

> Blog モデルの方に「見せて良いものの集合」を返す何かを実装しておいて、ここからはそれに対して find するだけ、という形が好ましいです〜！

ここの実装方法が全く思い浮かびませんでした。。。
Blog.hoge.find という形になるということでしょうか？

> なので、その意味でもなにかに対して find させて、見つからなければ自然と ActiveRecord::RecordNotFound が発生するように、というのが正攻法だったりします！（つまりここは出題者からの「find が raise する ActiveRecord::RecordNotFound を使ってくれぇ〜」というメッセージだったわけですね🙂）

ActiveRecord::RecordNotFound を使って欲しい意図は汲み取れたのですが、 現行のコードの Blog.find(params[:id]) は必ず成立するので、苦肉の策で手動で raise させてしまいました。。。
ActiveRecord 内に find された時に走らせる何かを指定できるのかと調査しましたがそれもなさそうでした。

Reply:
> ここの実装方法が全く思い浮かびませんでした。。。
> Blog.hoge.find という形になるということでしょうか？

まさにそういう形です！！ヒントを出すと、こういう仕組みが使えそうです: https://railsguides.jp/active_record_querying.html#%E3%82%B9%E3%82%B3%E3%83%BC%E3%83%97

Reply:
> スコープでは、where、joins、includesなど、これまでに登場したメソッドをすべて使えます。どのスコープメソッドも、常にActiveRecord::Relationオブジェクトを返します

ActiveRecord::Relation を返すんですね！だから、where で返ってきたオブジェクトに対して find メソッドを使用できるんですね。
勝手に配列を返してくると思い込んでいて、 Book.find か Book.find_by で RecordNotFound を引き起こすしかないと思っていました。

Reply:
修正しました！
https://github.com/yamatatsu10969/haunted-blog/pull/1/commits/78d9caa51648d5b2289ee3898b0000e1b96b34c8